tests/test_otel.py

```diff
@@ -0,0 +1,349 @@
+)
+@pytest.mark.parametrize(
+    ("headers", "tid", "sid", "pid", "ps"),
+    [
+        (None, ["trace_id"] * 4, ["span_id"] * 4, [None] * 4, ["0"] * 4),
+        (
+            context,
+            ["trace_id", trace_id] * 2,
+            ["span_id"] * 4,
+            [None, span_id] * 2,
+            ["0", "1"] * 2,
+        ),
+    ],
+    ids=["no context", "context"],
+)
+@pytest.mark.parametrize("path", ["/vars"], ids=["trace on"])
+def test_variables(
+    headers, path, tid, sid, pid, ps, idx, trace_service, client
+):
+    r = client.get(f"http://127.0.0.1:18080{path}", headers=headers)
+
+    span = trace_service.get_span()
+    assert span.name == path
+
+    assert r.headers["X-Otel-Trace-Id"] == decode_id(span, tid[idx])
+    assert r.headers["X-Otel-Span-Id"] == decode_id(span, sid[idx])
+    assert r.headers.get("X-Otel-Parent-Id") == pid[idx]
+    assert r.headers["X-Otel-Parent-Sampled"] == ps[idx]
+
+
+@pytest.mark.skip(reason="variables are present when trace is off")
+@pytest.mark.parametrize(
+    "headers", [None, context], ids=["no context", "context"]
+)
+@pytest.mark.parametrize("path", ["/notrace"], ids=["trace off"])
+def test_no_variables(headers, path, trace_service, client):
+    r = client.get(f"http://127.0.0.1:18080{path}", headers=headers)
+
+    time.sleep(0.01)  # wait for spans
+    assert len(trace_service.batches) == 0
+
+    assert r.headers.get("X-Otel-Trace-Id") is None
+    assert r.headers.get("X-Otel-Span-Id") is None
+    assert r.headers.get("X-Otel-Parent-Id") is None
+    assert r.headers.get("X-Otel-Parent-Sampled") is None
+
+
+@pytest.mark.parametrize(
+    ("headers", "idx"),
+    [(None, 0), (context, 1)],
+    ids=["no context", "context"],
+)
+@pytest.mark.parametrize(
+    ("path", "tparent", "tstate"),
+    [
+        (
+            "/context-ignore",
+            [None, context["Traceparent"]],
+            [None, context["Tracestate"]],
+        ),
+        (
+            "/context-extract",
+            [None, context["Traceparent"]],
+            [None, context["Tracestate"]],
+        ),
+        (
+            "/context-inject",
+            ["00-trace_id-span_id-01", "00-trace_id-span_id-01"],
+            [None, None],
+        ),
+        (
+            "/context-propagate",
+            ["00-trace_id-span_id-01", f"00-{trace_id}-span_id-01"],
+            [None, context["Tracestate"]],
+        ),
+    ],
+    ids=["ignore", "extract", "inject", "propagate"],
+)
+def test_trace_context(
+    headers, path, idx, tparent, tstate, trace_service, client
+):
+    r = client.get(f"http://127.0.0.1:18080{path}", headers=headers)
+
+    span = trace_service.get_span()
+    assert span.name == path
+
+    # Validate headers
+    assert r.headers.get("X-Otel-Traceparent") == (
+        "-".join(decode_id(span, val) for val in tparent[idx].split("-"))
+        if tparent[idx] is not None
+        else None
+    )
+    assert r.headers.get("X-Otel-Tracestate") == tstate[idx]
```

Comment:
Let's make it a bit easier to comprehend:

```suggestion
@pytest.mark.parametrize("parent", [None, parent_ctx])
@pytest.mark.parametrize("path",
    ["/ignore", "/extract", "/inject", "/propagate"])
def test_context(client, trace_service, parent, path):
    headers = trace_headers(parent)
    r = client.get(f"http://127.0.0.1:18080{path}", headers=headers)

    span = trace_service.get_span()

    if path in ["/extract", "/propagate"] and parent:
        assert span.trace_id.hex() == parent.trace_id
        assert span.parent_span_id.hex() == parent.span_id
        assert span.trace_state == parent.state

    if path in ["/inject", "/propagate"]:
        headers = trace_headers(
                    TraceContext(span.trace_id.hex(),
                                 span.span_id.hex(),
                                 span.trace_state or None))

    assert r.headers.get("X-Otel-Traceparent") == headers["Traceparent"]
    assert r.headers.get("X-Otel-Tracestate") == headers["Tracestate"]
```

Reply:
Applied